main/app_main.c

```diff
@@ -68,282 +62,444 @@ void app_main()
+    .method    = HTTP_GET,
+    .handler   = handle_jpg_stream,
+};
+
+static ip4_addr_t get_ip_addr(void)
+{
+    tcpip_adapter_ip_info_t ip_info; 
+   	
+    // IP address.
+    tcpip_adapter_get_ip_info(TCPIP_ADAPTER_IF_STA, &ip_info);
+
+    return ip_info.ip;
+}
 
-static esp_err_t event_handler(void *ctx, system_event_t *event)
+static httpd_handle_t start_webserver(void)
 {
-    switch (event->event_id) {
-        case SYSTEM_EVENT_STA_START:
-            esp_wifi_connect();
-            break;
-        case SYSTEM_EVENT_STA_GOT_IP:
-            xEventGroupSetBits(s_wifi_event_group, CONNECTED_BIT);
-            s_ip_addr = event->event_info.got_ip.ip_info.ip;
-            break;
-        case SYSTEM_EVENT_STA_DISCONNECTED:
-            esp_wifi_connect();
-            xEventGroupClearBits(s_wifi_event_group, CONNECTED_BIT);
-            break;
-        default:
-            break;
+    httpd_handle_t server = NULL;
+    httpd_config_t config = HTTPD_DEFAULT_CONFIG();
+
+    // Start the httpd server
+    ESP_LOGI(TAG, "Starting server on port: '%d'", config.server_port);
+    if (httpd_start(&server, &config) == ESP_OK) {
+        // Set URI handlers
+        ESP_LOGI(TAG, "Registering URI handlers");
+
+        sensor_t * sensor = esp_camera_sensor_get();
+
+        ip4_addr_t s_ip_addr = get_ip_addr();
+       
+        if (sensor->pixformat == PIXFORMAT_GRAYSCALE) {
+            httpd_register_uri_handler(server, &bmp);
+            httpd_register_uri_handler(server, &bmp_stream);
+            httpd_register_uri_handler(server, &pgm);
+
+            ESP_LOGI(TAG, "Open http://" IPSTR "/bmp for a single image/bmp gray image", IP2STR(&s_ip_addr));
+            ESP_LOGI(TAG, "Open http://" IPSTR "/bmp_stream for multipart/x-mixed-replace stream of gray bitmaps", IP2STR(&s_ip_addr));
+            ESP_LOGI(TAG, "Open http://" IPSTR "/pgm for a single image/x-portable-graymap image", IP2STR(&s_ip_addr));
+        } else if (sensor->pixformat == PIXFORMAT_RGB888) {
+            httpd_register_uri_handler(server, &bmp);
+            httpd_register_uri_handler(server, &bmp_stream);
+            
+            ESP_LOGI(TAG, "Open http://" IPSTR "/bmp for single image/bitmap image", IP2STR(&s_ip_addr));
+            ESP_LOGI(TAG, "Open http://" IPSTR "/bmp_stream for multipart/x-mixed-replace stream of bitmaps", IP2STR(&s_ip_addr));
+        } else if (sensor->pixformat == PIXFORMAT_JPEG) {
+            httpd_register_uri_handler(server, &jpg);  
+            httpd_register_uri_handler(server, &jpg_stream);          
+     
+            ESP_LOGI(TAG, "Open http://" IPSTR "/jpg for single image/jpg image", IP2STR(&s_ip_addr));
+            ESP_LOGI(TAG, "Open http://" IPSTR "/jpg_stream for multipart/x-mixed-replace stream of JPEGs", IP2STR(&s_ip_addr));
+        }
+
+        return server;
     }
-    return ESP_OK;
+
+    ESP_LOGI(TAG, "Error starting server!");
+    return NULL;
 }
 
-static void initialise_wifi(void)
```

Comment:
Removing this method and it call, removes the AP/STA support obviously

Reply:
I will do some research to try to add AP mode.

Reply:
Just separate your changes and the pure porting to idf4.1 (and try to do this according to the latest https://github.com/espressif/esp-idf-template, i.e. add CMakeLists.txt to be able to use idf.py) into multiple commits and pull-requests.